README.md
# Unattended Installer (Semi Automated) for OpenStack (LIBERTY)
Reynaldo R. Martínez P.
E-Mail: TigerLinux at Gmail dot com
Caracas, Venezuela.

## Introduction

This installer was made to automate the tasks of creating a virtualization infrastructure based on OpenStack. So far, There are two "flavors" for this installer: One for  Centos 7 and one for Ubuntu 14.04 LTS.

All two versions produce a fully production-usable OpenStack. You can use this installer to make a single-node all-in-one OpenStack server, or a more complex design with controller and compute nodes.

In summary, this installer can produce an OpenStack virtualization service completely usable in production environments, however, remember that the "bugs" factor don't depend solely on us. From time to time OpenStack packages can bring us some bugs too. We are using rpm/deb packages from Ubuntu and Redhat repositories and they can have their own bugs. 

## Using the Installer.

### First

* READ, READ, READ and after some rest, READ AGAIN!. *

Read everything you can from * OpenStack * if you want to venture into the virtualization in the cloud World. If you do not like reading, then support yourself on someone who can do the reading. Please do not try to use this Installer without having any knowledge at hand. View file `NOTES.txt` to understand a little more about the knowledge which you should have.

You can begin here: * http://docs.openstack.org *

The big world of * OpenStack * includes several technologies from the world of Open-source and the world of networks that must be understood thoroughly before even attempting any installation of OpenStack, whether you use this installation tool or any other. In short, if you do not have the knowledge, do not even try. Gain the knowledge first, then proceed.

Before using the installer, you must prepare your server of servers. Again, in the file `NOTES.txt` you will find important points that you should understand before start an installation using this tool. The installer will make some validations, should yield negative results, will abort the process.

### Second: Edit the installer main configuration file.

The installer has a central configuration file: `./configs/main-config.rc`. This file is well documented so * if you did your homework and studied * about * OpenStack *, you will know what to change there. There are very obvious things like passwords, IP addresses, modules to install and dns domains or domain names.

In the version by default, the configuration file has selections modules to install what is known as an "all-in-one" (an OpenStack monolithic service with controller-compute capabilites). You can just change the IP with the one assigned to your server (please DO NOT use * localhost * and DO NOT use a Dynamic DHCP assigned IP).

Additionally, there are some modules that are in default "no":

* Heat *
* Swift *
* Trove * 
* Sahara *
* SNMP *

We recommend to activate swift install option "only If you are really going to use it". * Swift * alone is almost as extensive as OpenStack. Use if you REALLY know what you're doing and if you are REALLY going to use it. 

The SNMP module installs monitoring variables useful if you want to monitor OpenStack with SNMP but does not install any monitoring application. The variables are described (if you install the support) in `/etc/snmp/snmpd.conf`.

NOTE: Files for ZABBIX agent in the "Goodies" directory are also included.

If you want to install an "all-in-one" only change passwords, IP addresses and mail domains and *dhcp/dnsmasq* info appearing in the configuration file.

After updating its configuration file, run at the root of directory script the following command:

```
# ./main-installer.sh install
```

The installer asks if you want to proceed (y/n).

If you run the installer with the additional parameter * auto *, it will run automatically without asking you confirmation. Example:

```
# ./main-installer.sh install auto
```

You can save all outputs produced by the installer using the tool `tee`. Example:

```bash
./main-installer.sh install | tee -a /var/log/my_log_de_install.log
```

## Controlling the installer behavior

As mentioned before, you can use this installer for more complex designs. Example:

* A single all-in-one monolithic server *
* A cloud with a controller-compute and several compute nodes *
* A cloud with a pure controller and several compute nodes *

### Controller node:

If your controller node will include a compute service (controller + compute, or an all-in-one server), the following variable in the configuration file must be set to “no”:

```bash
nova_without_compute="no"
```

If you use ceilometer in the controller, and likewise the controller includes compute service, the following variable must also be set to "no":

```bash
ceilometer_without_compute="no"
```

However, if you are installing a "pure" controller (without compute service) set the following variables to "yes":

```bash
nova_without_compute="yes"
ceilometer_without_compute="yes"
```

### Compute nodes:

For the compute nodes, you must set to "yes" (this is mandatory) the installation variables for Nova and Neutron modules. The remaining modules (glance, cinder, horizon, trove, sahara and heat) must be set to "no". If you are using Ceilometer in the controller, you also must set it's installation variable to “yes” along with the ones for Nova and Neutron. In Addition, the following variables in sections of nova and neutron must be set to "yes":

```bash
nova_in_compute_node="yes"
neutron_in_compute_node="yes"
```

And if you are using ceilometer also the following variable must be "yes" for compute nodes:

```bash
ceilometer_in_compute_node="yes"
```

You must place the IP's for the services running in the controller (neutron, keystone, glance and cinder) and the Ip's for the Database and message broker backends. This is valid for either a controller or a compute:

```bash
novahost="Controller IP Address"
glancehost="Controller IP Address"
cinderhost="Controller IP Address"
neutronhost="Controller IP Address"
keystonehost="Controller IP Address"
messagebrokerhost="Message Broker IP Address"
dbbackendhost="Database Server IP Address"
vncserver_controller_address | spiceserver_controller_address = "Controller IP Address"
```

If you use ceilometer, the same case applies:

```bash
ceilometerhost = "Controller IP Address"
```

For compute nodes, you must place the following variables with the IP in the compute node:

```bash
neutron_computehost = "Compute Host IP Address"
nova_computehost = "Compute Host IP Address "
```

### Database Backend

The installer has the ability to install and configure the database service, and also it will create all the databases. This is completely controllable by the configuration file through the following variables:

```bash
dbcreate = "yes"
dbinstall = "yes"
dbpopulate = "yes"
```

With these three options set to "yes", the database software is installed, will be configured and databases will be created using all the information contained in the configuration file.

> ** WARNING **: If you choose these options, you must ensure that there is
> NO database software previously installed or the process will fail.

By using our installation tool, you can choose to install and/or use between MySQL-based and PostgreSQL-based engines. In the specific case of Centos, we'll use MariaDB instead of MySQL.


If you prefer to “not install” any database software because you already have one installed somewhere else (a database farm), and also have the proper administrative access to the database engine, set the variables as follows:

```bash
dbcreate = "yes"
dbinstall = "no"
dbpopulate = "yes"
```

With this, the database software will not be installed, but it's up to you (or your * DBA *) to ensure you have full administrative access to create and modify databases in the selected backend.

If you do not want to install database software nor create databases (we assume that you already have previously created then in a farm or a separate server or even manually in the controller) set the three values "no":

```bash
dbcreate = "no"
dbinstall = "no"
dbpopulate = "no"
```

In any case, always remember to properly set the database-control variables inside the installer configuration file.


### RPC Messaging backend (Message Broker)

As part of the components to install and configure, the installer installs and configure the software for * AMQP * (the * Message Broker *). This step * IS * mandatory for a controller or * all-in-one * OpenStack server. If your server or servers have a message broker already installed, conflicts can occur that prevent the correct operation of the installation.

Again, the installer configuration file will control which AMPQ service to install and configure. You can choose from RabbitMQ and QPID. The installer will install and configure everything include the account and password that will be used by all OpenStack services to talk with the Message Broker Service.

NOTE: We strongly recommend to use RABBITMQ as message broker for OpenStack.


### Console Manager (NOVNC / SPICEHTML5)

Through a configurable option in the installer configuration file (consoleflavor), you can choose between NoVNC and SpiceHTML5. If you want to eventually use SSL for the Dashboard, please leave the default (novnc) as it easier to configure with SSL.


### Trove

If you choose to install trove, this installation tool will install and configure all the software needed, but IT WILL NOT configure trove-ready images. That's part of your tasks as a Cloud Administrator. Please follow recomendations from the community regarding proper-configured glance
images for trove. The "very big secret" of proper trove deployment is the glance-image. Fail to do that properlly, and forget about trove working
the way it should.

Tips for a properlly working trove image:

- Cloud init must be installed on the image and configured for start at boot time. Please eliminate "mounts" from /etc/cloud/cloud.cfg or your vm
  will try to auto-mount the ephemeral disk. This can interfere with trove guest agent activities.
- The trove guest agent MUST BE installed and configured in the image. Also, give sudo root-powers to the "trove" account on the glance image. The
  guest agent perform some changes in the vm that requires root access. The trove guest agent (if installed from ubuntu/centos repositories) uses
  a "trove" account to run.
- Install the database engine software in the glance image too. Trove guest agent can do this for you in many ways too.
- TRICK: You can install the guest agent, configure it, create the "sudo" permissions, and install the database software vía Cloud init. You just
  need to create a file /etc/trove/cloudinit/DATASTORE-NAME.cloudinit (sample: /etc/trove/cloudinit/mysql.cloudinit) with the commands needed to
  do everything. This file can be any script-based languaje (sh, bash, etc.).
- Flavors: If you plan to use locally-based storage for trove (instead of cinder-based), remember to choose a flavor for your database services
  that contains an ephemeral disk. Trove requires an extra disk for the database.


### Support Scripts installed with this solution

This installer will place a OpenStack Services control script in the “/usr/local/bin” path:

```bash
openstack-control.sh OPTION
```

The script uses the following options:

1. **enable**: Enables the services to start at boot time.
2. **disable**: disable services start at boot time.
3. **start**: starts all services.
4. **stop**: stops all services.
5. **restart**: restart all services.
6. **status**: displays the status of all services.

NOTE: We used or best judgment to ensure the proper start/stop order in the openstack-control.sh script. That being said, you could benefit a lot by using this script to control you cloud instead of the order normally set by “init”, “systemctl” or “upstart”. A good choice can be to place the script inside rc.local file. Your choice.

**IMPORTANT NOTE**: Again, We recommend using the openstack-control.sh script to initialize all OpenStack services!. Put all openstack services in "disable" state with "openstack-control.sh disable" and call the script with the "start" option from inside the /etc/rc.local file:

```bash
/usr/local/bin/openstack-control.sh start
```

This script is included by the installer in every single OpenStack node (controller and compute nodes)

You can also control individual OpenStack modules with the script:

```bash
/usr/local/bin/openstack-control.sh OPTION MODULE
```

Samples:

```bash
/usr/local/bin/openstack-control.sh start nova
```

```bash
/usr/local/bin/openstack-control.sh restart neutron
```

```bash
/usr/local/bin/openstack-control.sh status cinder
```

By the moment, we support the following modules:
- keystone
- swift
- glance
- cinder
- neutron
- nova
- ceilometer
- heat
- sahara



```bash
openstack-log-cleaner.sh
```

The installer will place a script “openstack-log-cleaner.sh” in the path “/usr/local/bin” that have the ability to “clean” all OpenStack related logs.

This script is called during the final phase of installation to clean all logs before leaving the server installed and running for the very first time, but can also be used by you “Cloud Administrator” to clean all OpenStack related logs whenever you consider it necessary.

```bash
compute-and-instances-full-report.sh
```

This script is also copied by the installer onto /usr/local/bin directory. The function of this script is give a report of all compute nodes in the openstack cloud and it's related virtual machines (instances) including the IP or IP's assigned to the instances.


### Keystone Environment Admin Variables

This installer will place the following files in your OpenStack Nodes:

```bash
/root/keystonerc_admin
/root/keystonerc_fulladmin
```

This files include your "admin" credentials (user/password included) along the URL endpoints for Keystone Service. The file first file use the normal public endpoint at port tcp 5000. The second one, uses the full admin port 35357.

Sourcing the *keystonerc_admin* file in your environment will allow you to perform normal administration tasks, not included the ones related to keystone advanced tasks. Sourcing the *keystonerc_fulladmin* file in your environment will give you "super cow god-like powers" over your cloud installation.

Then:

Normal admin tasks:

```bash
source /root/keystonerc_admin
```

Super-cow god-like powers:

```bash
source /root/keystonerc_fulladmin
```


### STARTING VIRTUAL MACHINES AVOIDING I/O STORMS

If you suffer a total blackout and your cloud service goes completely down, and then try to start it including all virtual machines (instances), chances are that you will suffer a I/O storm. That can easily collapses all your servers or at least slow them down for a while.

We include a script called “openstack-vm-boot-start.sh” that you can use to start all your OpenStack VM's (instances) with a little timeout between each virtual machine. You need to include the name or UUID of the instances that you want to start automatically in the following file:

```bash
/etc/openstack-control-script-config/nova-start-vms.conf
```

Place the script in the rc.local file ONLY in the controller node.

NOTE: The names of the VMs must be obtained from "nova list" command.


### DNSMASQ

Neutron dhcp-agent uses **DNSMASQ** for IP assignation to the VM's (instances). We include a customized dnsmasq-control file with some samples that you can use to fine-tune your dhcp-agent:

```
/etc/dnsmasq-neutron.d/neutron-dnsmasq-extra.conf
```

There are commented examples in the file. Use these examples to pass options to
different instances of dnsmasq created for each subnet where you select the option to use * dhcp *.

Recommendation: Try to have a good **DNS** structure for your cloud.


### Installer modularization

While the main setup process "* main-installer.sh *" is responsible for calling each module of each installer component, these modules are really independent of one another, to the point that they can be called sequentially and manually by you. Is not the common case, but can be done. The normal order of execution for each module is as follows (assuming that all components will be installed):

* requeriments.sh
* messagebrokerinstall.sh
* databaseinstall.sh
* requeriments-extras.sh (only present for Ubuntu Server 14.04 LTS)
* keystoneinstall.sh
* swiftinstall.sh
* glanceinstall.sh
* cinderinstall.sh
* neutroninstall.sh
* novainstall.sh
* ceilometerinstall.sh
* heatinstall.sh
* troveinstall.sh
* saharainstall.sh
* snmpinstall.sh
* horizoninstall.sh
* postinstall.sh


Then again, we do not recommend to run those modules out of the main installer, unless of course you know exactly what are you doing.


### RECOMMENDATIONS FOR INSTALLATION IN CENTOS AND UBUNTU SERVER.

#### Centos 7:

1. Install Centos with the selection of packages for "Infrastructure Server". Make sure you have properly installed and configured both SSH and NTP. Ntpdate is also recommended. Again, a proper DNS infrastructure is very recommended.

2. Add the EPEL and RDO repositories (see "NOTES.txt").

3. Install and configure OpenVSWitch (again, see "NOTES.txt").

*WARNING*: OpenStack does not support MySQL lower than 5.5. See notes and take proper steps. If you use our installation tool in order to install database support, we will install MariaDB 5.5.x directly obtained from RDO repositories.

IMPORTANT NOTE: The installer disables Centos 7 SELINUX. We had found some bugs, specially when using PostgreSQL and with some scenarios with NOVA-API.

#### Ubuntu 14.04 LTS:

1. Install Ubuntu Server 14.04 LTS standard way and select as an additional package "OpenSSH Server". Install and configure the ntpd service. Also SSH. It is also recommended to use ntpdate.

2. Install and configure OpenVSWitch (see "NOTES.txt").

As you can see in all cases, NTP and SSH are very important. Fail to configure those services correctly, and prepare to have a live full of misery.

#### Debian 8:

A debian 8 installer is in the works, but for now, we can't deliver dates until we test the overall quality of the OpenStack packages in the third-party repos we are going to use. Stay tuned for further anouncements !!.


### Cinder:

If you are using CINDER with lvm-iscsi, be sure to have a free partition or disk to create a LVM called "cinder-volumes". Example (free disk /dev/sdc):

```bash
pvcreate /dev/sdc
vgcreate cinder-volumes /dev/sdc
```

Another example with an free /dev/sda3 partition:

```bash
pvcreate /dev/sda3
vgcreate cinder-volumes /dev/sda3
```

### Swift:


DOCS/CHANGELOG.txt
1.0.12.ub1404lts - May 28, 2016

- Minor modifications in neutron install module.
- Added new script for cpu metrics report:
  "instance-cpu-metrics-report.sh".
  This script reports the min/max/avg cpu usage
  for every active instance in your cloud. This
  script requires ceilometer installed and fully
  active.

1.0.11.ub1404lts - May 18, 2016

- Corrected buggy sed call in ceilometer
  install module.

1.0.10.ub1404lts - May 17, 2016

- Added missing config item in glance registry.

1.0.9.ub1404lts - May 04, 2016

- Minor corrections on Ceilometer install module

1.0.8.ub1404lts - April 19, 2016

- Added and changed config items for ceilometer.
- Added ceilometer polling service.
- Added extra config items for cinder, nova, heat,
  glance and sahara regarding ceilometer.

1.0.7.ub1404lts - February 27, 2016

- Extra changes on trove install module regarding
  keystone region and swift access

1.0.6.ub1404lts - February 21, 2016

- BUG corrected in heat install module regarding
  proper authentication with Keystone V3.

1.0.5.ub1404lts - February 20, 2016

- Minor changes on trove install module and trove
  notes on main readme.

1.0.4.ub1404lts - February 18, 2016

- Minor correction in trove install module. The
  guest-agent config file was not property of the
  trove user.
- Added more notes regarding trove in the main
  Readme.

1.0.3.ub1404lts - February 16, 2016

- Extensive rework of trove installation module. New
  config options for trove in main config file, and
  new procedure included to make a "sample" trove guest
  agent config file in order to make easier for the
  cloud admin to create the glance trove images.
- Minor corrections in Ceilometer install module.
- Added missing config items in Neutron regarding
  notifications.

1.0.2.ub1404lts - February 10, 2016

- BUG corrected in cinder module. We had an
  incorrect option for glusterfs backend.
- BUG corrected in neutron module. The dns_domain
  option was wrongly placed into dhcp agent. The
  option is now where it should be (in neutron.conf)

1.0.1.ub1404lts - February 01, 2016

- Minor corrections on main readme (typos).
- Deleted duplicated entry in main config, for
  "neutron" configuration.
- Added extra checks in uninstall module.
- Added extra information regarding "bootable=1"
  bug in Horizon. See file:
  "ALERTS-DOMAIN-AND-BUGS-WITH-POSTGRES"
- New feature (developed by Ender Mujica from
  OpenStackVE Community): dhcp agent redundancy
  configurable on installer (see main config,
  "neutron" section) - Tksss U very much Ender !.
  With this new feature, we can activate extra
  dhcp agents on compute nodes.

1.0.0.ub1404lts - December 02, 2015

- First "ready for production" release.
- Completed "ALERT-AND-BUGS" documentation with
  some possible manual fixes for BUG affecting
  horizon operations when PostgreSQL is selected
  as database backend.
- Added extra documentation if "openstack-vlans.txt"
  file regarding new ways to configure interfaces
  for openvswitch in Ubuntu/Debian.

0.9.3.ub1404lts - November 17, 2015

- Added ALERT-AND-BUGS documentation to the installer
  regarding Domains usage with keystone V3 and some
  OpenStack bugs with postgres affecting cinder and
  trove.

0.9.2.ub1404lts - November 17, 2015

- Added Keystone V3 support to Horizon.
- Corrected some config items in neutron module.
- Included some ALERTS in the main config file.

0.9.1.ub1404lts - November 16, 2015

- Corrected some missing config items in neutron module.
- Added the capability to provision VLAN external net-
  works along the FLAT external ones.

0.9.0.ub1404lts - November 12, 2015

- All modules converted to LIBERTY from former installer
  made for Kilo.
- Fully converted to Identity V3.
- New options for FLAT networks configuration (neutron
  install module).


main-installer.sh
#!/bin/bash
#
# Unattended installer for OpenStack.
# Reynaldo R. Martinez P.
# E-Mail: TigerLinux@Gmail.com
#
# Main Installer Script
# Version: 1.0.12.ub1404lts "Nuclear Lynx"
# May 28, 2016
#

PATH=$PATH:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin

#
# This is our main installation script. It's function is basically read the main configuration file
# in "./configs" directory, source all installation variables, and then install every component in
# the right order. Normally the order is as follows:
# 1.- Basic python libraries and other base packages. Libvirt/qemu/kvm, sudo, etc.
# 2.- Message Broker (either rabbit or qpid - only if it's a controller).
# 3.- Database installation, configuration and provisioning (if it's a controller).
# 4.- Extra libraries (only on debian/ubuntu based installs).
# 5.- Keystone (if it's a controller, installs the sofwtare, in other cases, justo provision the en-
#     vironment files
# 6.- Application Modules (if apply and where apply) in order:
#     - Swift.
#     - Glance.
#     - Cinder.
#     - Neutron.
#     - Nova.
#     - Ceilometer.
#     - Heat.
#     - Trove.
#     - Sahara.
#     - Horizon.
# 7.- Basic SNMP Support.
# 8.- Post-install with maintenance crontabs and scripts
#
# This installation tool will allow you to construct an OpenStack cloud with the following possible
# configurations:
# - Monolitic ALL-IN-ONE OpenStack Server (basically an ALL-Services controller/compute).
# - Distributed Cloud with a Controller/Compute and several compute nodes.
# - Distributed Cloud with a Pure Controller and several compute nodes.
# - Fully distributed cloud with controller services distributed in many nodes, and several compute nodes
#
# The behaviour of this script is completelly controlled in the ./config/main-config.rc file. By properlly
# configuring this file, you can deploy your cloud the way you want.
#

case $1 in
"install")

	if [ -f ./configs/main-config.rc ]
	then
		source ./configs/main-config.rc
		mkdir -p /etc/openstack-control-script-config
		date > /etc/openstack-control-script-config/install-init-date-and-time
		chown -R root.root *
		find . -name "*" -type f -exec chmod 644 "{}" ";"
		find . -name "*.sh" -type f -exec chmod 755 "{}" ";"
	else
		echo "I can't access my own configuration"
		echo "Please check you are executing the installer in its correct directory"
		echo "Aborting !!!!."
		echo ""
		exit 0
	fi

	clear

	echo ""
	echo "OPENSTACK UNATTENDED INSTALLER"
	echo "Flavor: OpenStack LIBERTY for Ubuntu 14.04lts"
	echo "Made by: Reynaldo R. Martinez P."
	echo "E-Mail: TigerLinux@Gmail.com"
	echo "Version 1.0.12.ub1404lts \"Nuclear Lynx\" - May 28, 2016"
	echo ""
	echo "I'll verify all requiremens"
	echo "If any requirement is not met, I'll stop and inform what's missing"
	echo ""
	echo "Requirements"
	echo "- OS: Ubuntu Server 14.04 LTS x86_64"
	echo "- This script must be executed by root account (don't use sudo)"
	echo "- All ubuntu original repositories must be enabled and available"
	echo "- Ubuntu Cloud Archive OpenStack-LIBERTY repository must be enabled and available"
	echo "- OpenVSwitch must be installed and configured with at least br-int bridge"
	echo "- If you wish to install swift, the filesystem should be mounted in /srv/node"
	echo ""
	echo "NOTE: You can use the tee command if you want to log all installer actions. Example:"
	echo "./main-installer.sh install | tee -a /var/log/my_install_log.log"
	echo ""

	case $2 in
	auto|AUTO)
		echo "Automated mode activated. No additional questions will be made"
		;;
	*)
		echo -n "Do you wish to continue ? [y/n]:"
		read -n 1 answer
		echo ""
		case $answer in
		y|Y)
			echo ""
			echo "Starting verifications"
			echo ""
			;;
		*)
			echo ""
			echo "Aborting by user request !!!"
			echo ""
			exit 0
			;;
		esac
		;;
	esac


	echo ""
	echo "Installing requirements and performing more validations"
	echo ""

	./modules/requeriments.sh
	
	if [ -f /etc/openstack-control-script-config/libvirt-installed ]
	then
		echo ""
		echo "Requeriments installed and initial validations done"
		echo "I'll continue with the OpenStack modules installation"
		echo ""
	else
		echo ""
		echo "Something bad happened. Validations failed"
		echo "Aborting the installation !!"
		echo ""
		exit 0
	fi


	echo "Ready... let's continue working"
	echo ""

	rm -rf /tmp/keystone-signing-*
	rm -rf /tmp/cd_gen_*

	if [ $messagebrokerinstall == "yes" ]
	then
		echo ""
		echo "Installing message broker"
		./modules/messagebrokerinstall.sh
		
		if [ -f /etc/openstack-control-script-config/broker-installed ]
		then
			echo ""
			echo "Ready"
			echo ""
		else
			echo ""
			echo "Message broker installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	echo ""
	echo "Installing database support"
	echo ""

	./modules/databaseinstall.sh

	if [ -f /etc/openstack-control-script-config/db-installed ]
	then
		echo ""
		echo "Database support ready"
		echo ""
	else
		echo ""
		echo "Database support installation failed. Abroting !!"
		echo ""
		exit 0
	fi

	echo ""
	echo "Installing extra requeriments"
	echo ""
	
	./modules/requeriments-extras.sh

	if [ -f /etc/openstack-control-script-config/requeriments-extras-installed ]
	then
		echo ""
		echo "Extra requeriments ready. Continuing"
		echo ""
	else
		echo ""
		echo "Extra requirements installation failed. Aborting !!"
		echo ""
		exit 0
	fi


	if [ $keystoneinstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK KEYSTONE"

		./modules/keystoneinstall.sh

		if [ -f /etc/openstack-control-script-config/keystone-installed ]
		then
			echo "OPENSTACK KEYSTONE INSTALLED"
		else
			echo ""
			echo "Keystone installation failed. Aborting !!"
			echo ""
			exit 0
		fi

	else

		OS_URL="http://$keystonehost:35357/v3"
		OS_USERNAME=$keystoneadminuser
		OS_TENANT_NAME=$keystoneadminuser
		OS_PASSWORD=$keystoneadminpass
		OS_AUTH_URL="http://$keystonehost:5000/v3"
		OS_VOLUME_API_VERSION=2
		OS_PROJECT_DOMAIN_ID=$keystonedomain
		OS_USER_DOMAIN_ID=$keystonedomain
		OS_IDENTITY_API_VERSION=3

		echo "# export OS_URL=$SERVICE_ENDPOINT" > $keystone_admin_rc_file
		echo "# export OS_TOKEN=$SERVICE_TOKEN" >> $keystone_admin_rc_file
		echo "export OS_USERNAME=$OS_USERNAME" >> $keystone_admin_rc_file
		echo "export OS_PASSWORD=$OS_PASSWORD" >> $keystone_admin_rc_file
		echo "export OS_TENANT_NAME=$OS_TENANT_NAME" >> $keystone_admin_rc_file
		echo "export OS_PROJECT_NAME=$OS_TENANT_NAME" >> $keystone_admin_rc_file
		echo "export OS_AUTH_URL=$OS_AUTH_URL" >> $keystone_admin_rc_file
		echo "export OS_VOLUME_API_VERSION=2" >> $keystone_admin_rc_file
		echo "export OS_IDENTITY_API_VERSION=3" >> $keystone_admin_rc_file
		echo "export OS_PROJECT_DOMAIN_ID=$keystonedomain" >> $keystone_admin_rc_file 
		echo "export OS_USER_DOMAIN_ID=$keystonedomain" >> $keystone_admin_rc_file
		echo "export OS_AUTH_VERSION=3" >> $keystone_admin_rc_file
		echo "PS1='[\u@\h \W(keystone_admin)]\$ '" >> $keystone_admin_rc_file

        	OS_AUTH_URL_FULLADMIN="http://$keystonehost:35357/v3"

        	echo "# export OS_URL=$SERVICE_ENDPOINT" > $keystone_fulladmin_rc_file
        	echo "# export OS_TOKEN=$SERVICE_TOKEN" >> $keystone_fulladmin_rc_file
        	echo "export OS_USERNAME=$OS_USERNAME" >> $keystone_fulladmin_rc_file
        	echo "export OS_PASSWORD=$OS_PASSWORD" >> $keystone_fulladmin_rc_file
        	echo "export OS_TENANT_NAME=$OS_TENANT_NAME" >> $keystone_fulladmin_rc_file
        	echo "export OS_PROJECT_NAME=$OS_TENANT_NAME" >> $keystone_fulladmin_rc_file
        	echo "export OS_AUTH_URL=$OS_AUTH_URL_FULLADMIN" >> $keystone_fulladmin_rc_file
        	echo "export OS_VOLUME_API_VERSION=2" >> $keystone_fulladmin_rc_file
		echo "export OS_IDENTITY_API_VERSION=3" >> $keystone_fulladmin_rc_file
                echo "export OS_PROJECT_DOMAIN_ID=$keystonedomain" >> $keystone_fulladmin_rc_file
                echo "export OS_USER_DOMAIN_ID=$keystonedomain" >> $keystone_fulladmin_rc_file
		echo "export OS_AUTH_VERSION=3" >> $keystone_fulladmin_rc_file
        	echo "PS1='[\u@\h \W(keystone_fulladmin)]\$ '" >> $keystone_fulladmin_rc_file

		mkdir -p /etc/openstack-control-script-config
		date > /etc/openstack-control-script-config/keystone-installed
		date > /etc/openstack-control-script-config/keystone-extra-idents
	fi

	if [ $swiftinstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK SWIFT"

		./modules/swiftinstall.sh

		if [ -f /etc/openstack-control-script-config/swift-installed ]
		then
			echo "OPENSTACK SWIFT INSTALLED"
		else
			echo ""
			echo "Swift installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	if [ $glanceinstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK GLANCE"

		./modules/glanceinstall.sh

		if [ -f /etc/openstack-control-script-config/glance-installed ]
		then
			echo "OPENSTACK GLANCE INSTALLED"
		else
			echo ""
			echo "Glance installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	if [ $cinderinstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK CINDER"

		./modules/cinderinstall.sh

		if [ -f /etc/openstack-control-script-config/cinder-installed ]
		then
			echo "OPENSTACK CINDER INSTALLED"
		else
			echo ""
			echo "Cinder installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	if [ $neutroninstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK NEUTRON"

		./modules/neutroninstall.sh

		if [ -f /etc/openstack-control-script-config/neutron-installed ]
		then
			echo "OPENSTACK NEUTRON INSTALLED"
		else
			echo ""
			echo "Neutron installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	if [ $novainstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK NOVA"

		./modules/novainstall.sh

		if [ -f /etc/openstack-control-script-config/nova-installed ]
		then
			echo "OPENSTACK NOVA INSTALLED"
		else
			echo ""
			echo "Nova installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	if [ $ceilometerinstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK CEILOMETER"

		./modules/ceilometerinstall.sh

		if [ -f /etc/openstack-control-script-config/ceilometer-installed ]
		then
			echo "OPENSTACK CEILOMETER INSTALLED"
		else
			echo ""
			echo "Ceilometer installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

        if [ $heatinstall == "yes" ]
        then
                echo ""
                echo "Installing OPENSTACK HEAT"

                ./modules/heatinstall.sh

                if [ -f /etc/openstack-control-script-config/heat-installed ]
                then
                        echo "OPENSTACK HEAT INSTALLED"
                else
                        echo ""
                        echo "Heat installation failed. Aborting !!"
                        echo ""
                        exit 0
                fi
        fi

	if [ $troveinstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK TROVE"
		
		./modules/troveinstall.sh
		
		if [ -f /etc/openstack-control-script-config/trove-installed ]
		then
			echo "OPENSTACK TROVE INSTALLED"
		else
			echo ""
			echo "Trove installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

        if [ $saharainstall == "yes" ]
        then
                echo ""
                echo "Installing OPENSTACK SAHARA"
		
                ./modules/saharainstall.sh
		
                if [ -f /etc/openstack-control-script-config/sahara-installed ]
                then
                        echo "OPENSTACK SAHARA INSTALLED"
                else
                        echo ""
                        echo "Sahara installation failed. Aborting !!"
                        echo ""
                        exit 0
                fi
        fi

	if [ $snmpinstall == "yes" ]
	then
		echo ""
		echo "INSTALLING SNMP SUPPORT"

		./modules/snmpinstall.sh

		if [ -f /etc/openstack-control-script-config/snmp-installed ]
		then
			echo "SNMP SUPPORT INSTALLED"
		else
			echo ""
			echo "SNMP Support failed to install, but because this is NOT critical,"
			echo "we will continue the installation"
			echo ""
		fi
	fi

	if [ $horizoninstall == "yes" ]
	then
		echo ""
		echo "Installing OPENSTACK HORIZON"

		./modules/horizoninstall.sh

		if [ -f /etc/openstack-control-script-config/horizon-installed ]
		then
			echo "OPENSTACK HORIZON INSTALLED"
		else
			echo ""
			echo "Horizon installation failed. Aborting !!"
			echo ""
			exit 0
		fi
	fi

	echo ""
	echo "Executing Post Install"
	./modules/postinstall.sh

	date > /etc/openstack-control-script-config/install-end-date-and-time

	echo ""
	echo "OPENSTACK INSTALLATION FINISHED"
	echo ""
	

	;;
"uninstall")

	if [ -f ./configs/main-config.rc ]
	then
		source ./configs/main-config.rc
	else
                echo "I can't access my own configuration"
                echo "Please check you are executing the installer in its correct directory"
                echo "Aborting !!!!."
                echo ""
		exit 0
	fi

	echo ""
	echo "All openstack related content will be erased from this server"
	echo ""
	case $2 in
	auto|AUTO)
		echo "Automated mode activated"
		;;
	*)
		echo -n "Are you sure you want to continue ? [y/n]:"
		read -n 1 answer
		case $answer in
		y|Y)
			echo ""
			echo "Uninstalling OpenStack services, packages and configurations"
			echo ""
			;;
		*)
			echo ""
			echo "Aborted by user request"
			echo ""
			exit 0
			;;
		esac
	esac
	./modules/uninstall.sh
	;;
*)
	echo ""
	echo "Usage: $0 install | uninstall [auto]"
	echo ""
	;;
esac

